### PR TITLE
Phase 5: Reynolds Scaling Augmentation — Physics-Informed Synthetic Data (8 parallel)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Generate **synthetic training samples at different Reynolds numbers** by applying boundary layer similarity scaling laws to existing samples. For attached flows on airfoils, the surface pressure coefficient Cp is approximately Re-independent, while the velocity field scales with Re. This means we can create physically plausible synthetic samples by: (1) keeping Cp constant, (2) scaling velocity to match the new Re.

**Why this should work:**
- val_ood_re (p_re=24.7) tests Re=4.445M — entirely outside our training Re range (~0.8-1.5M). Physics tells us Cp is nearly Re-independent for attached laminar flow.
- Synthetic Re augmentation creates training signal at Re values closer to the validation range without any external solver.
- From Blasius similarity: velocity profiles scale as u/U ~ f(y*sqrt(Re)), so synthetic samples at different Re can be generated by: `Ux_new = Ux * (Re_new/Re_orig)^alpha` where alpha ≈ 0.5.
- This directly targets p_re, our hardest single-Re metric.

## Instructions

Add a `--re_scaling_aug` flag. During training, with probability p_aug:
1. Pick a random training sample
2. Generate Re_new = Re_orig * uniform(0.5, 3.0)
3. Scale velocity: `y_new[:,:,0:2] = y[:,:,0:2] * (Re_new/Re_orig)**alpha`
4. Keep pressure unchanged (Cp is Re-independent): `y_new[:,:,2] = y[:,:,2]`
5. Update the Re feature in x: `x[:,:,11] = log(Re_new)`

```python
def re_scaling_augment(x, y, alpha=0.5, p_aug=0.3):
    if torch.rand(1).item() > p_aug:
        return x, y
    re_orig = x[:, 0, 11].exp()  # log_Re → Re
    re_scale = torch.empty(x.shape[0], 1, 1, device=x.device).uniform_(0.5, 3.0)
    re_new = re_orig.unsqueeze(-1).unsqueeze(-1) * re_scale
    vel_scale = (re_scale) ** alpha  # Blasius scaling
    y_aug = y.clone()
    y_aug[:, :, 0:2] = y[:, :, 0:2] * vel_scale  # scale Ux, Uy
    # pressure stays the same (Cp is Re-independent)
    x_aug = x.clone()
    x_aug[:, :, 11:12] = re_new.log().expand_as(x[:, :, 11:12])
    return x_aug, y_aug
```

### Hyperparameters across 8 GPUs:

| GPU | re_aug | alpha | p_aug | re_range | Notes |
|-----|--------|-------|-------|----------|-------|
| 0 | True | 0.5 | 0.3 | [0.5, 3.0] | Default Blasius |
| 1 | True | 0.3 | 0.3 | [0.5, 3.0] | Weaker scaling |
| 2 | True | 0.5 | 0.5 | [0.5, 3.0] | Higher aug probability |
| 3 | True | 0.5 | 0.3 | [0.3, 5.0] | Wider Re range |
| 4 | True | 0.5 | 0.3 | [1.0, 5.0] | Only scale UP (toward val Re) |
| 5 | True | 0.0 | 0.3 | [0.5, 3.0] | No velocity scaling (just Re perturbation) |
| 6 | True | 0.5 | 0.3 | [0.5, 3.0] | seed 43 |
| 7 | — | — | — | — | **Baseline** |

### Training commands:
```bash
BASE="--agent violet --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --seed 42"

CUDA_VISIBLE_DEVICES=0 python train.py $BASE --wandb_name "violet/p5-reaug-05-03" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.3
# GPU 7: Baseline
CUDA_VISIBLE_DEVICES=7 python train.py $BASE --wandb_name "violet/p5-baseline" --wandb_group phase5-re-aug
```

## Baseline
| Metric | Value |
|--------|-------|
| val/loss | 0.401 ± 0.005 |
| p_in | 12.95 ± 0.3 |
| p_oodc | 8.40 ± 0.4 |
| p_tan | 33.8 ± 0.5 |
| p_re | 24.7 ± 0.2 |

### Reproduce baseline:
```bash
python train.py --agent violet --wandb_name "violet/baseline" \
  --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp \
  --n_layers 3 --slice_num 96 --tandem_ramp \
  --domain_layernorm --domain_velhead --ema_decay 0.999 \
  --weight_decay 5e-5 --cosine_T_max 180 --disable_pcgrad \
  --pressure_first --pressure_deep
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT noam:

## Results

All 8 runs completed at 180-minute timeout (epoch 158–159). Peak memory: ~36.3 GB per GPU.

### Results Table

| GPU | Experiment | W&B Run | Epoch | val/loss | p_in | p_tan | p_oodc | p_re |
|-----|-----------|---------|-------|----------|------|-------|--------|------|
| 0 | α=0.5, p=0.3, [0.5,3.0] (default) | anwp1j32 | 159 | 0.4249 | 14.0 | 35.1 | 9.7 | 25.4 |
| 1 | α=0.3, p=0.3, [0.5,3.0] (weaker α) | upp5s1v0 | 159 | 0.4141 | 13.4 | 33.6 | 9.9 | 25.1 |
| 2 | α=0.5, p=0.5, [0.5,3.0] (higher p) | jf6drd1w | 159 | 0.4423 | 16.3 | 35.0 | 12.3 | 26.3 |
| 3 | α=0.5, p=0.3, [0.3,5.0] (wider range) | t373i33s | 159 | 0.4227 | 13.8 | 35.0 | 10.3 | 25.8 |
| 4 | α=0.5, p=0.3, [1.0,5.0] (scale-up only) | ffs0k8a0 | 159 | 0.4179 | 13.8 | 34.5 | 10.2 | 25.7 |
| 5 | α=0.0, p=0.3, [0.5,3.0] (no vel scaling) | re8s2tdc | 158 | 0.4011 | 12.6 | 32.7 | 8.6 | 24.9 |
| 6 | α=0.5, p=0.3, [0.5,3.0], seed=43 | h1vvmg74 | 158 | 0.4236 | 14.3 | 34.4 | 10.2 | 25.7 |
| **7** | **Baseline (no Re aug)** | **ty1flwma** | **159** | **0.3945** | **12.5** | **32.7** | **8.3** | **24.5** |

**Baseline from PR** (reproduced internally): val/loss 0.401 ± 0.005, p_in 12.95, p_oodc 8.40, p_tan 33.8, p_re 24.7

### Commands used

```bash
BASE="--agent noam --field_decoder --adaln_output --use_lion --lr 2e-4 --cosine_T_max 180 --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 --disable_pcgrad --pressure_first --pressure_deep --seed 42"

(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py $BASE --wandb_name "noam/p5-reaug-05-03" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.3 --re_range_lo 0.5 --re_range_hi 3.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=1 python train.py $BASE --wandb_name "noam/p5-reaug-03-03" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.3 --re_aug_prob 0.3 --re_range_lo 0.5 --re_range_hi 3.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=2 python train.py $BASE --wandb_name "noam/p5-reaug-05-05" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.5 --re_range_lo 0.5 --re_range_hi 3.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=3 python train.py $BASE --wandb_name "noam/p5-reaug-wide" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.3 --re_range_lo 0.3 --re_range_hi 5.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=4 python train.py $BASE --wandb_name "noam/p5-reaug-up" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.3 --re_range_lo 1.0 --re_range_hi 5.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=5 python train.py $BASE --wandb_name "noam/p5-reaug-alpha0" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.0 --re_aug_prob 0.3 --re_range_lo 0.5 --re_range_hi 3.0)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=6 python train.py $BASE --wandb_name "noam/p5-reaug-s43" --wandb_group phase5-re-aug --re_scaling_aug --re_alpha 0.5 --re_aug_prob 0.3 --re_range_lo 0.5 --re_range_hi 3.0 --seed 43)
(cd /workspace/senpai/cfd_tandemfoil && CUDA_VISIBLE_DEVICES=7 python train.py $BASE --wandb_name "noam/p5-baseline" --wandb_group phase5-re-aug)
```

### What happened

**Reynolds scaling augmentation did not help — it hurt performance across the board.**

The in-run baseline (GPU7, ty1flwma) achieved val/loss=0.3945, p_in=12.5, p_oodc=8.3, p_re=24.5 — all better than or equal to every Re-augmented variant. This is a clean negative result:

- **Best Re-aug result** (GPU1, α=0.3): p_re=25.1 vs baseline p_re=24.5. Still 0.6 points worse.
- **GPU5 (α=0.0, no velocity scaling)** matched the baseline on p_re (24.9 vs 24.5) and nearly matched on all metrics. This is the only Re-aug variant that didn't clearly hurt. However it also didn't help — with α=0.0, the augmentation reduces to pure Re-jitter in x without any y change, so the model just sees inconsistent (x, y) pairs.
- **GPU2 (p=0.5, more augmentation)** was the worst: p_in degraded from 12.5→16.3, p_oodc from 8.3→12.3. More synthetic data made things worse.
- **Wider/scale-up ranges** (GPU3, GPU4) did not help p_re either (25.7–25.8 vs 24.5 baseline).
- Seed-42 vs seed-43 replicate (GPU0 vs GPU6) produced similar results (p_re=25.4 vs 25.7), confirming the augmentation effect is consistent, not noise.

**Why it failed:** The Blasius scaling assumption `u ~ Re^alpha` is a rough laminar boundary layer approximation. The actual velocity field in our CFD data (which includes separation, recirculation, tandem interference) does not follow this simple scaling. The synthetic samples created corrupted (x, y) pairs — the features suggest one Re but the velocity magnitudes suggest another — adding noise rather than signal. The val_ood_re split tests Re=4.445M, well beyond even our augmented range [0.5x, 5x] of training Re, so the augmentation doesn't close the gap anyway.

The GPU7 baseline run (ty1flwma) with val/loss=0.3945 is slightly better than the PR's stated baseline of 0.401 — this is within the ±0.005 variance range.

### Implementation note: Feature index correction
The PR pseudocode used `x[:, :, 11]` for log_Re. The actual index is **13** (pos=2, saf=2, dsdf=8, is_surface=1, **then** log_Re at index 13). This was verified from `data/prepare_multi.py` and the implementation uses index 13. The PR's pseudocode appears to have a typo.

### Suggested follow-ups
- **p_re gap is structural, not data starvation.** At Re=4.445M the flow physics (turbulent transition, separation characteristics) may genuinely differ. A physics-informed boundary condition or Re-conditional normalization might help more than synthetic augmentation.
- **If Re augmentation is revisited**, keeping Cp fixed and scaling velocity is physically valid only when there's no separation. A mask that applies augmentation only to samples with small surface variance (attached flow indicator) might give cleaner synthetic data.
- **GPU5 result is interesting**: setting α=0 (Re-jitter in x only, no y change) nearly matched baseline. This confirms the y-scaling is what hurts — the Re feature jitter alone is harmless. Worth trying Re-jitter with very small p_aug as a regularizer.
